### PR TITLE
perf: improve job selection performance in case of potential ambiguity that is resolved by comprehensive ruleorder statements.

### PR DESCRIPTION
**Pull request by johanneskoester**
### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

### CODE REVIEW

**Comment by johanneskoester** [comment]
@dkuzminov you probably want to have a look at this?

File: snakemake/dag.py [hunk 1]
```diff
@@ -735,6 +735,11 @@ def update(
         exceptions = list()
         cycles = list()
 
+        # check if all potential producers are strictly ordered
+        jobs = sorted(jobs, reverse=True)
+        primary_job = jobs[0]
+        is_strictly_ordered = all(primary_job > job for job in jobs)
```
Comment (dkuzminov) [inline snakemake/dag.py:741]: This is not correct, as the first element shall be compared to the rest of the elements:
```python
is_strictly_ordered = all(primary_job > job for job in jobs[1:])
```

**Review by dkuzminov — CHANGES_REQUESTED** [review]
Overall this improvement is incorrect. Assume there is a job `primary` that dominates all other jobs. Next, imagine that this rule is disregarded later because of missing input. We would get an incorrect `is_strictly_ordered` value because we know nothing about the ordering of the rest of the jobs. We would select the first producer even if it is ambiguous.

File: snakemake/dag.py (hunk 1 above)
Comment (johanneskoester) [reply snakemake/dag.py]: good catch!

**Comment by johanneskoester** [comment]
> Overall this improvement is incorrect. Assume there is a job `primary` that dominates all other jobs. Next, imagine that this rule is disregarded later because of missing input. We would get an incorrect `is_strictly_ordered` value because we know nothing about the ordering of the rest of the jobs. We would select the first producer (other than `primary`) even if it is ambiguous.

Right, I missed that case indeed. I have worked a bit on it. Still not 100% validated in my mind, but I think this is more correct (although it is 20pm here now, so I might miss something again ;-).

**Comment by johanneskoester** [comment]
I think it is fine.

**Pull request merged**